File: app/src/main/java/com/example/advancedprayertimes/ui/PrayerTimeGraphicView.java
package com.example.advancedprayertimes.UI;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;

import com.example.advancedprayertimes.Logic.Entities.PrayerTimeEntity;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;

public class PrayerTimeGraphicView extends View
{
    private static final int backgroundRectangleColor = Color.argb(255, 35,41,53);
    private static Paint backgroundRectanglePaint = new Paint()
    {
        {
            setColor(backgroundRectangleColor);
            setAntiAlias(true);
        }
    };

    private static final int innerBackgroundRectangleColor = Color.rgb(0, 204, 102);
    private static Paint innerBackgroundRectanglePaint = new Paint()
    {
        {
            setAntiAlias(true);
            setColor(innerBackgroundRectangleColor);
        }
    };

    private static final int prayerTimeMainTextColor = Color.YELLOW;
    Paint textPaint = new Paint()
    {
        {
            setAntiAlias(true);
            setTextSize(60);
            setColor(prayerTimeMainTextColor);
        }
    };

    Paint prayerTimeTextPaint = new Paint()
    {
        {
            setAntiAlias(true);
            setTextSize(40);
            setColor(prayerTimeMainTextColor);
        }
    };

    private static final int currentTimeTextColor = Color.RED;
    Paint currentTimeTextPaint = new Paint()
    {
        {
            setAntiAlias(true);
            setTextSize(40);
            setColor(currentTimeTextColor);
        }
    };

    Paint indicatorRectanglePaint = new Paint()
    {
        {
            setAntiAlias(true);
            setColor(currentTimeIndicatorLineColor);
        }
    };

    private static final int currentTimeIndicatorLineColor = Color.WHITE;

    public static DateTimeFormatter dateFormat = DateTimeFormatter.ofPattern("HH:mm");
    private PrayerTimeEntity _displayPrayerTimeEntity = null;

    public PrayerTimeGraphicView(Context context, AttributeSet attrs)
    {
        super(context, attrs);

        this.setBackgroundColor(backgroundRectangleColor);
    }

    @Override
    protected void onDraw(Canvas canvas)
    {
        super.onDraw(canvas);

        if(this.getDisplayPrayerEntity() == null || this.getDisplayPrayerEntity().getDuration() == 0)
        {
            return;
        }

        //TODO: Move rectangle instantation outside of draw

        // INNER BACKGROUND RECTANGLE
        RectF innerBackgroundRectangle = new RectF(180, 110, this.getRight() - 100, 550);
        canvas.drawRoundRect(innerBackgroundRectangle, 20, 20, innerBackgroundRectanglePaint);

        // PRAYER TIME TEXT
        canvas.drawText(this.getDisplayPrayerEntity().getTitle(), (this.getWidth()/2)-50, 80, textPaint);

        // PRAYER TIME BEGINNING TEXT
        canvas.drawText(dateFormat.format(this.getDisplayPrayerEntity().getBeginningTime()), 60, 125, prayerTimeTextPaint);

        // PRAYER TIME END TEXT
        canvas.drawText(dateFormat.format(this.getDisplayPrayerEntity().getEndTime()), 60, innerBackgroundRectangle.height() + 120, prayerTimeTextPaint);

        drawCurrentTime(canvas, innerBackgroundRectangle);
    }

    private void drawCurrentTime(Canvas canvas, RectF innerBackgroundRectangle)
    {
        LocalDateTime currentDate = LocalDateTime.now();

        long timeShare = 0;

        // TODO: Fix Isha
        if(this.getDisplayPrayerEntity().getEndTime().isBefore(this.getDisplayPrayerEntity().getBeginningTime()) && currentDate.isBefore(this.getDisplayPrayerEntity().getBeginningTime()))
        {
            timeShare = ChronoUnit.MILLIS.between(this.getDisplayPrayerEntity().getBeginningTime().minusDays(1), currentDate);
        }
        else
        {
            timeShare = ChronoUnit.MILLIS.between(this.getDisplayPrayerEntity().getBeginningTime(), currentDate);
        }

        double percentage = (double) timeShare/this.getDisplayPrayerEntity().getDuration();

        int relativePos = (int) (innerBackgroundRectangle.height() * percentage);

        int calculatedTop = (int)innerBackgroundRectangle.top + relativePos;
        int calculatedBottom = calculatedTop + 2;

        Rect indicatorRectangle = new Rect((int)innerBackgroundRectangle.left, calculatedTop, (int)innerBackgroundRectangle.right, calculatedBottom);
        canvas.drawRect(indicatorRectangle, indicatorRectanglePaint);

        // CURRENT TIME TEXT
        canvas.drawText(currentDate.format(dateFormat), 60, indicatorRectangle.top + 16, currentTimeTextPaint);
    }

    public PrayerTimeEntity getDisplayPrayerEntity()
    {
        return _displayPrayerTimeEntity;
    }

    public void setDisplayPrayerEntity(PrayerTimeEntity displayPrayerTimeEntity)
    {
        _displayPrayerTimeEntity = displayPrayerTimeEntity;
    }
}

File: app/src/main/kotlin/com/example/advancedprayertimes/ui/PrayerTimeGraphicView.kt
package com.example.advancedprayertimes.ui

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import com.example.advancedprayertimes.logic.PrayerTimeEntity
import com.example.advancedprayertimes.logic.enums.EPrayerTimeType
import com.example.advancedprayertimes.logic.extensions.toStringByFormat
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.temporal.ChronoUnit

class PrayerTimeGraphicView(context: Context?, attrs: AttributeSet?) : View(context, attrs) {

    companion object {

        private val MAIN_BACKGROUND_COLOR = Color.argb(255, 35, 41, 53)
        private val PRAYER_TIME_COLOR = Color.rgb(124,197,107)

        private val PRAYER_MAIN_TEXT_COLOR = Color.YELLOW
        private val CURRENT_TIME_TEXT_COLOR = Color.RED
        private val CURRENT_TIME_INDICATOR_COLOR = Color.BLACK
        private val PRAYER_SUB_TIME_BACKGROUND_COLOR = Color.rgb(90,187,71)
        private val PRAYER_SUB_TIME_BORDER_COLOR = Color.WHITE
        private val PRAYER_SUB_TIME_TEXT_COLOR = Color.CYAN
    }

    private var innerBackgroundRectangle: RectF =
        RectF(180F, 110F, this.right - 100F, 550F)

    init {
        setBackgroundColor(MAIN_BACKGROUND_COLOR)
    }

    // region Paint Objects

    private var mainRectanglePaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            textSize = 60f
            color = PRAYER_TIME_COLOR
            style = Style.FILL_AND_STROKE
            strokeWidth = 3F
        }
    }

    private var prayerNameTextPaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            textSize = 60f
            color = PRAYER_MAIN_TEXT_COLOR
        }
    }

    private var prayerSubTimeNameTextPaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            textSize = 40f
            color = PRAYER_SUB_TIME_TEXT_COLOR
        }
    }

    private var prayerTimeBeginningEndTimeTextPaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            textSize = 40f
            color = PRAYER_MAIN_TEXT_COLOR
        }
    }

    private var currentTimeTextPaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            textSize = 40f
            color = CURRENT_TIME_TEXT_COLOR
        }
    }

    private var indicatorRectanglePaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            color = CURRENT_TIME_INDICATOR_COLOR
            style = Style.STROKE
            strokeWidth = 3F
        }
    }

    private var prayerSubTimeRectangleFillPaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            color = PRAYER_SUB_TIME_BACKGROUND_COLOR
            style = Style.FILL_AND_STROKE
            strokeWidth = 3F
        }
    }

    private var prayerSubTimeRectangleBorderPaint: Paint = object : Paint() {
        init {
            isAntiAlias = true
            color = PRAYER_SUB_TIME_BORDER_COLOR
            style = Style.STROKE
            strokeWidth = 1F
        }
    }

    // endregion Paint Objects

    var displayPrayerEntity: PrayerTimeEntity? = null

    override fun onDraw(canvas: Canvas) {

        super.onDraw(canvas)
        innerBackgroundRectangle.right = this.right - 100F

        if (displayPrayerEntity == null || displayPrayerEntity!!.durationMS == 0L) {
            return
        }

        drawBaseInformation(canvas)
    }

    private fun drawAsrSubtimes(canvas: Canvas) {

        // FIRST THIRD OF NIGHT
        val innerSubtimeOneBackgroundRectangle =
            RectF(
                innerBackgroundRectangle.width()/2,
                innerBackgroundRectangle.top,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.subtime1EndTime!!, innerBackgroundRectangle)
            )
        canvas.drawRoundRect(innerSubtimeOneBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeOneBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("Ikhtiyar", 675F, 185F, prayerSubTimeNameTextPaint)

        // SECOND THIRD OF NIGHT
        val innerSubtimeTwoBackgroundRectangle =
            RectF(
                innerSubtimeOneBackgroundRectangle.left,
                innerSubtimeOneBackgroundRectangle.bottom,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.subtime2EndTime!!, innerBackgroundRectangle)
            )
        canvas.drawRoundRect(innerSubtimeTwoBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeTwoBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("Normal", 675F, 340F, prayerSubTimeNameTextPaint)

        // THIRD THIRD OF NIGHT
        val innerSubtimeThreeBackgroundRectangle =
            RectF(
                innerSubtimeOneBackgroundRectangle.left,
                innerSubtimeTwoBackgroundRectangle.bottom,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.endTime!!, innerBackgroundRectangle).toFloat()
            )
        canvas.drawRoundRect(innerSubtimeThreeBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeThreeBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("Karaha", 675F, 500F, prayerSubTimeNameTextPaint)
    }

    private fun drawMaghribSubtimes(canvas: Canvas) {

        // FIRST THIRD OF NIGHT
        val innerSubtimeOneBackgroundRectangle =
            RectF(
                innerBackgroundRectangle.width()/2,
                innerBackgroundRectangle.top,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.subtime1EndTime!!, innerBackgroundRectangle)
            )
        canvas.drawRoundRect(innerSubtimeOneBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeOneBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("Normal", 675F, 285F, prayerSubTimeNameTextPaint)

        // THIRD THIRD OF NIGHT
        val innerSubtimeThreeBackgroundRectangle =
            RectF(
                innerSubtimeOneBackgroundRectangle.left,
                innerSubtimeOneBackgroundRectangle.bottom,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.endTime!!, innerBackgroundRectangle).toFloat()
            )
        canvas.drawRoundRect(innerSubtimeThreeBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeThreeBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("Karaha", 675F, 500F, prayerSubTimeNameTextPaint)
    }

    private fun drawIshaSubtimes(canvas: Canvas) {

        // FIRST THIRD OF NIGHT
        val innerSubtimeOneBackgroundRectangle =
            RectF(
                innerBackgroundRectangle.width()/2,
                innerBackgroundRectangle.top,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.subtime1EndTime!!, innerBackgroundRectangle)
            )
        canvas.drawRoundRect(innerSubtimeOneBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeOneBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("1/3", 555F, 185F, prayerSubTimeNameTextPaint)

        // SECOND THIRD OF NIGHT
        val innerSubtimeTwoBackgroundRectangle =
            RectF(
                innerSubtimeOneBackgroundRectangle.left,
                innerSubtimeOneBackgroundRectangle.bottom,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.subtime2EndTime!!, innerBackgroundRectangle)
            )
        canvas.drawRoundRect(innerSubtimeTwoBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeTwoBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("2/3", 555F, 322F, prayerSubTimeNameTextPaint)

        // THIRD THIRD OF NIGHT
        val innerSubtimeThreeBackgroundRectangle =
            RectF(
                innerSubtimeOneBackgroundRectangle.left,
                innerSubtimeTwoBackgroundRectangle.bottom,
                innerBackgroundRectangle.right,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.endTime!!, innerBackgroundRectangle).toFloat()
            )
        canvas.drawRoundRect(innerSubtimeThreeBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeThreeBackgroundRectangle, prayerSubTimeRectangleBorderPaint)
        canvas.drawText("3/3", 555F, 485F, prayerSubTimeNameTextPaint)

        // FIRST HALF OF NIGHT
        val innerSubtimeFourBackgroundRectangle =
            RectF(
                300 + innerSubtimeOneBackgroundRectangle.left,
                innerBackgroundRectangle.top,
                this.right - 100F,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.subtime3EndTime!!, innerBackgroundRectangle)
            )
        canvas.drawRoundRect(innerSubtimeFourBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeFourBackgroundRectangle, prayerSubTimeRectangleBorderPaint);
        canvas.drawText("1/2", 860F, 220F, prayerSubTimeNameTextPaint)

        // SECOND HALF OF NIGHT
        val innerSubtimeFiveBackgroundRectangle =
            RectF(
                300 + innerSubtimeOneBackgroundRectangle.left,
                innerSubtimeFourBackgroundRectangle.bottom,
                this.right - 100F,
                innerBackgroundRectangle.top + getRelativeDepthByTime(displayPrayerEntity!!.endTime!!, innerBackgroundRectangle)
            )
        canvas.drawRoundRect(innerSubtimeFiveBackgroundRectangle, 0f, 0f, prayerSubTimeRectangleFillPaint)
        canvas.drawRect(innerSubtimeFiveBackgroundRectangle, prayerSubTimeRectangleBorderPaint);
        canvas.drawText("2/2", 860F, 440F, prayerSubTimeNameTextPaint)
    }

    private fun drawBaseInformation(canvas: Canvas) {

        mainRectanglePaint.color = PRAYER_TIME_COLOR
        canvas.drawRoundRect(innerBackgroundRectangle, 20f, 20f, mainRectanglePaint)
        canvas.drawRoundRect(innerBackgroundRectangle, 20f, 20f, mainRectanglePaint)

        drawPrayerTimeTexts(canvas)

        if(displayPrayerEntity!!.prayerTimeType == EPrayerTimeType.Isha) {
            drawIshaSubtimes(canvas)
        } else if (displayPrayerEntity!!.prayerTimeType == EPrayerTimeType.Maghrib) {
            drawMaghribSubtimes(canvas)
        }else if (displayPrayerEntity!!.prayerTimeType == EPrayerTimeType.Asr) {
            drawAsrSubtimes(canvas)
        }

        drawCurrentTimeIndicator(canvas, innerBackgroundRectangle)
    }

    private fun drawPrayerTimeTexts(canvas: Canvas) {

        // PRAYER NAME TEXT
        canvas.drawText(
            displayPrayerEntity!!.title,
            (this.width / 2 - 50).toFloat(),
            80f,
            prayerNameTextPaint
        )

        // PRAYER TIME BEGINNING TEXT
        canvas.drawText(
            displayPrayerEntity!!.beginningTime!!.toStringByFormat("HH:mm"),
            60f,
            125f,
            prayerTimeBeginningEndTimeTextPaint
        )

        // PRAYER TIME END TEXT
        canvas.drawText(
            displayPrayerEntity!!.endTime!!.toStringByFormat("HH:mm"),
            60f,
            560f,
            prayerTimeBeginningEndTimeTextPaint
        )
    }

    private fun drawCurrentTimeIndicator(canvas: Canvas, innerBackgroundRectangle: RectF) {
        val currentDate = LocalDateTime.now()

        if(displayPrayerEntity?.durationMS == null) { return }

        val relativePos = getRelativeDepthByTime(LocalTime.now(), innerBackgroundRectangle) ?: return
        val calculatedTop = innerBackgroundRectangle.top.toInt() + relativePos
        val calculatedBottom = calculatedTop + 2

        val indicatorRectangle = Rect(
            innerBackgroundRectangle.left.toInt(),
            calculatedTop,
            innerBackgroundRectangle.right.toInt(),
            calculatedBottom
        )
        canvas.drawRect(indicatorRectangle, indicatorRectanglePaint)

        // CURRENT TIME TEXT
        canvas.drawText(
            currentDate.toStringByFormat("HH:mm"),
            60f,
            (indicatorRectangle.top + 16).toFloat(),
            currentTimeTextPaint
        )
    }

    private fun getRelativeDepthByTime(time: LocalTime, innerBackgroundRectangle: RectF): Int {

        val stuff = if (displayPrayerEntity!!.endTime!!.atDate(LocalDate.now()).isBefore(displayPrayerEntity!!.beginningTime!!.atDate(LocalDate.now()))
            &&
            time.isBefore(displayPrayerEntity!!.beginningTime)
        ) {
            ChronoUnit.MILLIS.between(displayPrayerEntity!!.beginningTime!!.atDate(LocalDate.now()).minusDays(1), time.atDate(LocalDate.now()))
        } else {
            ChronoUnit.MILLIS.between(displayPrayerEntity!!.beginningTime, time)
        }

        val percentage = stuff.toDouble() / displayPrayerEntity!!.durationMS!!
        return (innerBackgroundRectangle.height() * percentage).toInt()
    }
}